crates/primitives/src/prune/limiter.rs

```diff
@@ -0,0 +1,123 @@
+    start: Instant,
+}
+
+impl PruneTimeLimit {
+    fn new(limit: Duration) -> Self {
+        Self { limit, start: Instant::now() }
+    }
+
+    fn is_limit_reached(&self) -> bool {
+        self.start.elapsed() >= self.limit
+    }
+}
+
+impl PruneLimiter {
+    /// Sets the limit on the number of deleted entries (rows in the database).
+    /// If the limit was already set, it will be updated and the number of already deleted entries
+    /// will not be reset.
+    pub fn set_deleted_entries_limit(mut self, limit: usize) -> Self {
+        if let Some(deleted_entries_limit) = self.deleted_entries_limit.as_mut() {
+            deleted_entries_limit.limit = limit;
+        } else {
+            self.deleted_entries_limit = Some(PruneDeletedEntriesLimit::new(limit));
+        }
+
+        self
+    }
+
+    /// Sets the limit on the number of deleted entries (rows in the database) to a biggest
+    /// multiple of the given denominator that is smaller than the existing limit.
+    ///
+    /// If the limit wasn't set, does nothing.
+    pub fn floor_deleted_entries_limit_to_multiple_of(mut self, denominator: NonZeroUsize) -> Self {
+        if let Some(deleted_entries_limit) = self.deleted_entries_limit.as_mut() {
+            deleted_entries_limit.limit =
+                (deleted_entries_limit.limit / denominator) * denominator.get();
+        }
+
+        self
+    }
+
+    /// Returns `true` if the limit on the number of deleted entries (rows in the database) is
+    /// reached.
+    pub fn is_deleted_entries_limit_reached(&self) -> bool {
+        self.deleted_entries_limit.as_ref().map_or(false, |limit| limit.is_limit_reached())
+    }
+
+    /// Increments the number of deleted entries by the given number.
+    pub fn increment_deleted_entries_count_by(&mut self, entries: usize) {
+        if let Some(limit) = self.deleted_entries_limit.as_mut() {
+            limit.deleted += entries;
+        }
+    }
+
+    /// Increments the number of deleted entries by one.
+    pub fn increment_deleted_entries_count(&mut self) {
+        self.increment_deleted_entries_count_by(1)
+    }
+
+    /// Returns the number of deleted entries left before the limit is reached.
+    pub fn deleted_entries_limit_left(&self) -> Option<usize> {
+        self.deleted_entries_limit.as_ref().map(|limit| limit.limit - limit.deleted)
+    }
+
+    /// Returns the limit on the number of deleted entries (rows in the database).
+    pub fn deleted_entries_limit(&self) -> Option<usize> {
+        self.deleted_entries_limit.as_ref().map(|limit| limit.limit)
+    }
+
+    /// Sets the time limit.
+    pub fn set_time_limit(mut self, limit: Duration) -> Self {
+        self.time_limit = Some(PruneTimeLimit::new(limit));
+
+        self
+    }
+
+    /// Returns `true` if time limit is reached.
+    pub fn is_time_limit_reached(&self) -> bool {
+        self.time_limit.as_ref().map_or(false, |limit| limit.is_limit_reached())
+    }
```

Comment:
hmm why the false positive?

Reply:
because if pruning exists because of some other reason, i.e. `self.start.elapsed() >= self.limit` is false, and then some more work is done, **then** the check is run again, it might be true the second time as more time will have elapsed.

Reply:
> because if pruning exists because of some other reason, i.e. self.start.elapsed() >= self.limit is false, and then some more work is done, then the check is run again, it might be true the second time as more time will have elapsed.

I don't understand why it matters; if the check has already returned "true" on the first call (due to entries limit), so on the second call even though another limit (time) has been reached, it still returns "true"?

Reply:
yeah but if it returns false the first time, and then true, you have a false positive

Reply:
I'm talking ab if someone in the future decides to use that method on the limiter to check if pruning timed out